### ISSUE TITLE
Double rendering (Lifecycle hook scheduled a cascading update) after item selection

### ISSUE THREAD

**Issue by eldargab** [downshift-js/downshift#686]
Hello!

When using Downshift with controlled `selectedItem` property it always triggers cascading update from `componentDidUpdate()` after selection has changed.

Here is a guilty excerpt 
```js
if (
  this.isControlledProp('selectedItem') &&
  this.props.selectedItemChanged(
    prevProps.selectedItem,
    this.props.selectedItem,
  )
) {
  this.internalSetState({
    type: stateChangeTypes.controlledPropUpdatedSelectedItem,
    inputValue: this.props.itemToString(this.props.selectedItem),
  })
}
```

There should be a way to move this piece of code away from `componentDidUpdate` hook to run before any rendering occurs.

**Comment by silviuavram** [comment]
Hello. I don't think it should. Maybe it depends on how you are using it. What does a cascading update mean? Can you provide more information?

**Comment by eldargab** [comment]
Cascading update is the following situation:

1. Component received new props/state and update was scheduled
2. Re-rendering occurred and all changes were committed do DOM
3. Immediately after that (typically from `componentDidUpdate()`) component receives new props/state again which causes another update cycle.

From efficiency point of view it's much better to apply/derive all state changes at once and not to trigger additional update cycles.

React detects such situation and issues performance warning "Lifecycle hook scheduled a cascading update" which can be seen from Chrome profiler.

Bellow is a screenshot of profiling of official [typeahead](https://zn8k2x793x.codesandbox.io/) example. Selection event is shown.

<img width="1552" alt="Screenshot 2019-03-12 at 09 00 22" src="https://user-images.githubusercontent.com/762157/54178043-6b2e4d80-44a5-11e9-95b5-51d65835b8a3.png">


There is two problems with this example.

1. Above mentioned `.setState()` from `componentDidUpdate()`.
2. `parent.setState()` from `onStateChange()` hook, which is also called after DOM updates, not before.

One should be able to mitigate problem 2 by moving `onStateChange()` logic to `stateReducer()`, but problem 1 is unfixable without upstream modifications.

**Comment by silviuavram** [comment]
still on vacation, so I cannot do any real debugging, but you can create a PR with improvements and I will take a look after I come back. Thanks!

**Comment by decafdennis** [comment]
There is a second reason triggering select/change handlers in response to state changes is problematic: if one of these event handlers throws an exception, it will kill the React rendering pipeline. On the other hand, typical event handlers are out of the rendering pipeline and will print errors to the console but not kill React.

For example, we had a bug in an `onSelect` handler, which is called from a `setState` completion callback (at a similar time as `componentDidUpdate`) and it caused a full React crash.

**Comment by yuriploc** [comment]
@silviuavram, see if this https://github.com/reduxjs/react-redux/issues/834 issue helps you.

**Comment by silviuavram** [comment]
merged https://github.com/downshift-js/downshift/pull/841 to stop re-rendering the items (if the items are pure components) when their highlighted status does not change. However I don't know if it affects this issue.

**Comment by eldargab** [comment]
> merged #841 to stop re-rendering the items (if the items are pure components) when their highlighted status does not change. However I don't know if it affects this issue.

@silviuavram, no, it doesn't. The issue is still there. You should stop doing

```
this.setState(newState, callbackToNotifyParentAboutEvents)
```

to resolve it.
